File: training-data/cpp/adapter_pattern.cpp
#include <iostream>
#include <memory>

class ModernPrinter {
public:
    virtual void printText(const std::string& text) = 0;
    virtual ~ModernPrinter() = default;
};

class LegacyPrinter {
public:
    void oldPrint(const char* text) { std::cout << "legacy: " << text << std::endl; }
};

class PrinterAdapter : public ModernPrinter {
public:
    explicit PrinterAdapter(std::unique_ptr<LegacyPrinter> legacy) : legacy(std::move(legacy)) {}
    void printText(const std::string& text) override { legacy->oldPrint(text.c_str()); }

private:
    std::unique_ptr<LegacyPrinter> legacy;
};

int main() {
    std::unique_ptr<ModernPrinter> printer = std::make_unique<PrinterAdapter>(std::make_unique<LegacyPrinter>());
    printer->printText("hello adapter");
    return 0;
}

File: training-data/swift/adapter_pattern.swift
protocol ModernPrinter {
    func print(text: String)
}

// Incompatible legacy interface we cannot change.
final class LegacyPrinter {
    func legacyPrint(_ payload: String) {
        Swift.print("[legacy] \(payload)")
    }
}

// Adapter makes LegacyPrinter conform to the interface client code expects.
final class LegacyPrinterAdapter: ModernPrinter {
    private let legacy: LegacyPrinter

    init(_ legacy: LegacyPrinter) {
        self.legacy = legacy
    }

    func print(text: String) {
        legacy.legacyPrint(text)
    }
}

final class NativePrinter: ModernPrinter {
    func print(text: String) {
        Swift.print("[native] \(text)")
    }
}

func run(printer: ModernPrinter, message: String) {
    printer.print(text: message)
}

let printers: [ModernPrinter] = [
    NativePrinter(),
    LegacyPrinterAdapter(LegacyPrinter()),
]

for printer in printers {
    run(printer: printer, message: "hello")
}
